### ISSUE TITLE
Certificates from system keystore are ignored by kubernetes client

### ISSUE THREAD

**Issue by ibuziuk** [fabric8io/kubernetes-client#711]
There are two possible ways for providing certificates while creating OpenShift / K8S client:

- `KUBERNETES_CERTS_CA_FILE` - using the **ENV VAR** and creating client using default constructor
- `withCaCertFile()` - providing path to the certificate file if client is created via` ConfigBuilder`

However, if certificate is already installed in the system, client will ignore it and would expect certificate to be explicitly set during  creation [1]

It looks like the kubernetes client creates an empty keystore and install the provided certificate into it.
Basically, this approach ignores system keystore and valid certificates from it are simply not taken into account.

The issue should be reproducible for everyone who uses k8-client to talk to a different cluster, as it seems to assume only this clusters key is valid and ignore all other root chains.

[1] https://github.com/redhat-developer/che-starter/blob/master/src/main/java/io/fabric8/che/starter/openshift/OpenShiftClientWrapper.java#L64

**Comment by ethanfrogers** [comment]
I'm dealing with a similar issue. From within a pod, I'm not able to communicate with the cluster using the cert from the ServiceAccount. The cert is self-signed.

**Comment by bartoszmajsak** [comment]
I cannot assign, but @matousjobanek is working on it.

**Comment by ibuziuk** [comment]
@bartoszmajsak awesome!!

**Comment by bartoszmajsak** [comment]
@iocanel can you assign @matousjobanek to this issue?

**Comment by MatousJobanek** [comment]
Hi,
I think that this is caused by the fact that there is no file that would be loaded when the KeyStore is created:
https://github.com/fabric8io/kubernetes-client/blob/master/kubernetes-client/src/main/java/io/fabric8/kubernetes/client/internal/CertUtils.java#L68
If I try (locally) to change the logic to:
~~~java
    char[] passphrase = "secrets".toCharArray();
    trustStore.load(new FileInputStream("/home/mjobanek/.keystore"),passphrase);
~~~
then it works.

Similar for the keystore: https://github.com/fabric8io/kubernetes-client/blob/master/kubernetes-client/src/main/java/io/fabric8/kubernetes/client/internal/CertUtils.java#L100

@iocanel could you shed more light on this, please? Is there any other way to load the truststore/keystore? If not, would be possible to load a default truststore/keystore file or some other file that would be specified using a parameter?

**Comment by iocanel** [comment]
I think that the config object should allow the user to specify trustore/keystore file or fallback to some sensible default (?). @jimmidyson: thoughts?

**Comment by jimmidyson** [comment]
I think it makes sense to load the default keystore and add custom certificates to that as specified from config. That keystore should be overridable as well though as you suggest.

**Comment by MatousJobanek** [comment]
I've sent a proposal how the issue could be solved: https://github.com/fabric8io/kubernetes-client/pull/738
@ibuziuk is it what you had expected?
@iocanel are you fine with the implementation or is there anything that I'm missing. I can also try to implement some tests or additional functionality of specifying custom keystore file/passphrase if you want.

**Comment by ibuziuk** [comment]
@matousjobanek I think that using default "changeit" will not work for many cases e.g.

```
 echo "Change password to system keystore for ${OSO_DOMAIN_NAME}"
 keytool -storepasswd -new $KEYSTORE_PASSWORD -keystore $JAVA_HOME/lib/security/cacerts -storepass changeit
```
Maybe it would be better to fall back on some ENV_VAR for it ?

Also, from API perspective maybe it would be better to have some methods which will give more control e.g.

> loadJavaKeyStore() / loadUserKeyStore()


CC: @aslakknutsen @hectorj2f WDYT?

**Comment by MatousJobanek** [comment]
> echo "Change password to system keystore for ${OSO_DOMAIN_NAME}"
> keytool -storepasswd -new $KEYSTORE_PASSWORD -keystore $JAVA_HOME/lib/security/cacerts -storepass changeit

Where I can find this code ^

> Maybe it would be better to fall back on some ENV_VAR for it ?
> Also, from API perspective maybe it would be better to have some methods which will give more control e.g.

Ok, I can implement it. I just thought that you had expected that the files (on default locations) should be loaded automatically (which can by done only using default passphrase).
Adding env parameters for specifying a custom file and passphrase was intended as the next step.
I can work only with the information that is written in this issue and based on it assume what you would need.

So to clarify the task/issue:
You don't expect any automatic functionality of loading default keystores. You need a ENV_VAR (or API method) for specifying which keystore file and passprase should be used for loading trusted certificates....? Am I right?

**Comment by ibuziuk** [comment]
> echo "Change password to system keystore for ${OSO_DOMAIN_NAME}"
> keytool -storepasswd -new $KEYSTORE_PASSWORD -keystore $JAVA_HOME/lib/security/cacerts -storepass changeit
> Where I can find this code ^

@hectorj2f could you please point to the script which installs certificate for keycloak (this initially was done also for che-starter, but than it was removed[1] because java keystore was not really used by k8s client) 

> `Ok, I can implement it. I just thought that you had expected that the files (on default locations) should be loaded automatically (which can by done only using default passphrase).`

Does it make sense  to load files from default locations using custom storepass if it was specified? e.g. fall back on `changeit` but if env var is specified use it as storepass ?


[1] https://github.com/redhat-developer/che-starter/commit/ee5acbc64b0e2bc70b713435fddc47547e583b3e#diff-f04f8a8a54296b988abbe80fb177d50dL34

**Comment by bartoszmajsak** [comment]
It's based on InstallCert https://github.com/almighty/keycloak/blob/f0daf238d59146b86a1676e100c6e8e4f3eeef40/docker/install_certificate.sh

**Comment by MatousJobanek** [comment]
> could you please point to the script which installs certificate for keycloak (this initially was done also for che-starter, but than it was removed[1] because java keystore was not really used by k8s client)

I think that you meant this one: https://github.com/redhat-developer/che-starter/blob/master/docker-entrypoint.sh#L27

> Does it make sense to load files from default locations using custom storepass if it was specified? e.g. fall back on changeit but if env var is specified use it as storepass ?

I think that it does. But you will need to specify the env var on your own. So, you will require additional logic that will detect cases when the passphrase is the default one and when not...

**Comment by MatousJobanek** [comment]
Looking at the install_certificate.sh file, yeah then it's not a problematic at all. 
Ok, I'll change the PR so it checks also the KEYSTORE_PASSWORD property... May I consider this property as appropriate one - that it doesn't follow the fabric8 property convention format: https://github.com/fabric8io/kubernetes-client#configuring-the-client

**Comment by ibuziuk** [comment]
@matousjobanek well, I think it should have the same convention format as other fabric8 properties -
 https://github.com/fabric8io/kubernetes-client#configuring-the-client

It's just a matter of setting it in our codebase properly

**Comment by hectorj2f** [comment]
It was already mentioned in the thread but we did it in two different ways:
https://github.com/redhat-developer/che-starter/blob/master/docker-entrypoint.sh
and
https://github.com/almighty/keycloak/blob/master/docker/install_certificate.sh

**Comment by MatousJobanek** [comment]
Ok, I changed the PR - I introduced new properties:
* kubernetes.truststore.file / KUBERNETES_TRUSTSTORE_FILE
* kubernetes.truststore.passphrase / KUBERNETES_TRUSTSTORE_PASSPHRASE
* kubernetes.keystore.file / KUBERNETES_KEYSTORE_FILE
* kubernetes.keystore.passphrase / KUBERNETES_KEYSTORE_PASSPHRASE

to make the setting of TrustStore/KeyStore file and passphrase customizable.
If no file is specified, then it tries to load files on default locations:
~/.keystore for KeyStore
java_jre_path/lib/security/cacerts for TrustStore
if the passphrase is not specified using the corresponding property, then the default one "changeit" is used.
I've added also corresponding API methods into the builder.

I hope that this is better and will work for you.
Tomorrow I'll write some tests for it.

**Comment by MatousJobanek** [comment]
The tests are in the PR, I've verified it also locally so it should work. 
If you want to use the default trustStore file, then just set KUBERNETES_TRUSTSTORE_PASSPHRASE to your secrets and the file should be loaded.
Any other concerns/comments? @ibuziuk @hectorj2f @iocanel

**Comment by rohanKanojia** [comment]
@MatousJobanek : Polite ping, Is there something left here or everything was covered up in #738

**Comment by MatousJobanek** [comment]
@rohanKanojia AFAIK everything was covered, so we can close this issue
